libbpf-cargo/src/gen/btf.rs

```diff
@@ -601,9 +607,10 @@ impl<'s> GenBtf<'s> {
             // Set `offset` to end of current var
             offset = datasec_var.offset + datasec_var.size as u32;
 
+            let ptr = if indirect { "*mut " } else { "" };
```

Comment:
do I understand correctly that because of this mut pointer, any access to this data will have to be unsafe?

If my understanding is correct, then wouldn't it be better to handle it exactly like what libbpf-cargo does for .rodata/.data/.bss sections? There is not conceptual difference here, beyond that .data/.bss memory is mmap()'ed, while for struct_ops it's just allocated with malloc() by libbpf. For Rust side it's just some memory that we can access directly.

I might be missing some complexities, but it's not clear from the patch itself, sorry.

Reply:
> do I understand correctly that because of this mut pointer, any access to this data will have to be unsafe?

Yes, that's correct.

> If my understanding is correct, then wouldn't it be better to handle it exactly like what libbpf-cargo does for .rodata/.data/.bss sections?

You mean by generating a bunch of getters for the individual members? I don't know how much value that adds, realistically. I find the existing generation infrastructure to be overly complicated and over engineered, with strange dependencies across crate boundaries, confusing naming (in my opinion), and entire classes of types that I see no real use for, though admittedly I may be missing some nuances as well (I have plans to look into improving that). Half of it does not map clearly to Rust anyway, because ownership of any of this data is not in the Rust realm, resulting in issues such those mutability conflicts we "fixed" earlier https://github.com/libbpf/libbpf-rs/issues/606. And then there are chances that what is contained one level down in the struct is a pointer anyway, so you'll end up using `unsafe` regardless.

If it's "just" about getters for the first level structs as opposed to having pointers, we can probably add that. Let me know if that's what you mean.

Reply:
Yep, that's what I meant.

So let's look at what we have right now for tcp_ca skeleton:

```
        #[derive(Debug, Copy, Clone)]
        #[repr(C)]
        pub struct bss {
            pub ca_cnt: i32,
        }
        #[derive(Debug, Copy, Clone)]
        #[repr(C)]
        pub struct struct_ops {
            pub ca_update: *mut tcp_congestion_ops,
        }
        #[derive(Debug, Copy, Clone)]
        #[repr(C)]
        pub struct tcp_congestion_ops {
            pub ssthresh: *mut std::ffi::c_void,
            pub cong_avoid: *mut std::ffi::c_void,
            pub set_state: *mut std::ffi::c_void,
            pub cwnd_event: *mut std::ffi::c_void,
            pub in_ack_event: *mut std::ffi::c_void,
            pub pkts_acked: *mut std::ffi::c_void,
            pub min_tso_segs: *mut std::ffi::c_void,
            pub cong_control: *mut std::ffi::c_void,
            pub undo_cwnd: *mut std::ffi::c_void,
            pub sndbuf_expand: *mut std::ffi::c_void,
            pub get_info: *mut std::ffi::c_void,
            pub name: [i8; 16],
            pub owner: *mut std::ffi::c_void,
            pub list: list_head,
            pub key: u32,
            pub flags: u32,
            pub init: *mut std::ffi::c_void,
            pub release: *mut std::ffi::c_void,
            pub __pad_152: [u8; 40],
        }
```

First, `pub cwnd_event: *mut std::ffi::c_void` and other callbacks should be exposed as getter/setter for `Option<&mut libbpf_rs::Program>` (not a `void *` raw pointer), so that it's possible to get/set (or reset, i.e., set to NULL at low-level of things) callbacks to various BPF programs using normal Rust code.

For
```
       pub struct struct_ops {
            pub ca_update: *mut tcp_congestion_ops,
        }
```
we should probably have ca_update as a method that returns a reference to tcp_congestion_ops type, not a raw pointer. Think about ca_update as conceptually the same thing as .bss map:

```
pub struct bss {
            pub ca_cnt: i32,
}
```

We have 

```
pub fn bss(&self) -> &tcp_ca_types::bss { ... }
pub fn bss_mut(&mut self) -> &mut tcp_ca_types::bss {... }
```

I think we should have the same for each struct_ops map itself. Given we expose it as direct `repr(C)`, I'd remove function pointers from it and would leave only data members (again, just like for bss), so user can directly read/write only data fields, while accessing (and setting) program references through dedicated getter/setter generated for each callback within struct_ops structure (could probably be done as methods on tcp_congestion_ops type, while leaving data members as pub fields?).

All the above is not something that *theoretically* is necessary. It's what sched_ext is either already using in C code (data members setting) or is going to use (setting BPF program references, I don't think they've updated the code to do this, because it just landed in libbpf very recently) to handle old kernels and changing interfaces in sched_ext's struct_ops. We support that on C side, and so the goal here is a feature parity on Rust side.

Just doing raw pointer manipulation is a good escape hatch for some advanced corner case scenarios, but what I'm describing above (setting a field, setting/resetting callback program) is going to be a typical use case for struct_ops going forward, so we should try to provide best possible and easy to use interface.

I'm sure there are Rust language limitations preventing us from implementing it exactly how I outlined above, but hopefully I explained the overall capabilities that are required.

Reply:
Sounds good. Will take care of that. Somehow I stumbled over a bunch of issues https://github.com/libbpf/libbpf-rs/pull/728 https://github.com/libbpf/libbpf-rs/pull/729 when building the `tcp_ca` example on a different computer that I had to fix before getting things building again.

Reply:
As discussed in person, we can't really work with `&Program` references as there are inherent borrowing conflicts when setting one on one originating in one part of the skeleton in another. I adjusted the code to work with `bpf_program` pointers instead of `std::ffi::c_void` and added getters that eliminate one pointer access. Setting programs seems to work fine as per the example.